### PR TITLE
Issue #30 - Factories can now be automatically built…

### PR DESCRIPTION
**Pull request by smendez-axway**
…based on configuration stored in key/value pairs in Properties object

### CODE REVIEW

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                    if (parameterTypes.length == 1) {
+                        String propertyName = method.getName().substring(3);
+                        Object value = getPropertyValue(parameterTypes[0], properties, baseName, propertyName);
+                        if (value != null) { // do not call setter if value is null
+                            method.invoke(builder, value);
+                        } else if (isMandatory(method)) {
+                            throw new ConfigurationException("Parameter '" + propertyName + "' for '"+ componentBuilderClazz.getClass().getCanonicalName()+"' is mandatory!");
+                        }
+                    } else {
+                        LOG.error("The configured builder has setters with more than one parameter {{}.{}}", componentBuilderClazz.getName(), method.getName());
```
Comment (xfournet) [inline iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java]: why not an exception ?

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    @Nullable
+    private <T> T getPropertyValue(Class<T> clazz, Properties properties, String baseName, String propertyName) {
+        if (clazz.isAssignableFrom(Properties.class)) {
+            return extractProperties(clazz, properties, baseName);
+        } else {
+            String fullName = baseName + "." + camelCaseToLowerCaseUnderscore(propertyName);
+            if (clazz.isAssignableFrom(Path.class)) {
+                return extractPath(clazz, properties, fullName);
```
Comment (xfournet) [inline iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java]: you can factorize out of individual if branch/extract methods:
* the getProperty
* the clazz.cast

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            Supplier<?> builder = constructor.newInstance();
+
+            for (Method method : componentBuilderClazz.getDeclaredMethods()) {
+                if (method.getName().startsWith("set")) {
+                    Class<?>[] parameterTypes = method.getParameterTypes();
+                    if (parameterTypes.length == 1) {
+                        String propertyName = method.getName().substring(3);
+                        Object value = getPropertyValue(parameterTypes[0], properties, baseName, propertyName);
+                        if (value != null) { // do not call setter if value is null
+                            method.invoke(builder, value);
```
Comment (xfournet) [inline iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java]: so we can't set the FileStoreFactoryBuilder "limitedSize" to null with that ?

File: iron-core/src/main/java/io/axway/iron/core/spi/file/FileSnapshotStoreFactoryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package io.axway.iron.core.spi.file;
+
+import java.nio.file.Path;
+import java.util.function.*;
+import javax.annotation.*;
+import io.axway.iron.spi.storage.SnapshotStoreFactory;
+
+public class FileSnapshotStoreFactoryBuilder implements Supplier<SnapshotStoreFactory> {
+    private Path m_dir;
+    private Integer m_limitedSize = 20;
```
Comment (xfournet) [inline iron-core/src/main/java/io/axway/iron/core/spi/file/FileSnapshotStoreFactoryBuilder.java]: better name for that ? eg transactionIdPaddingLenght or a better proposition ?

Comment (smendez-axway) [reply iron-core/src/main/java/io/axway/iron/core/spi/file/FileSnapshotStoreFactoryBuilder.java]: transactionIdPaddingLength it is now

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java (hunk 1 above)
Comment (smendez-axway) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java]: Initially I added the log so I can debug. But you're right, I'll throw an exception!

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java (hunk 2 above)
Comment (smendez-axway) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java]: ok.
I factorized the getProperty, and completely removed the clazz.cast

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java (hunk 3 above)
Comment (smendez-axway) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerFactoryBuilderConfigurator.java]: Good catch!
Now, a key without a value is considered as a set(null), and an absent key will keep the defaut value (if present)

**Pull request merged**